pandas/_libs/tslibs/offsets.pyx

```diff
@@ -304,6 +304,15 @@ class _BaseOffset(object):
     _day_opt = None
     _attributes = frozenset(['n', 'normalize'])
 
+    def __init__(self, n=1, normalize=False):
+        n = self._validate_n(n)
+        object.__setattr__(self, "n", n)
+        object.__setattr__(self, "normalize", normalize)
+        object.__setattr__(self, "_cache", {})
```

Comment:
what is _cache for?

Reply:
If we don't create this explicitly in `__init__` then `@cache_readonly` lookups will try to create it, which will raise.

Reply:
can you put this on the class? (or is that really weird)?

Reply:
```
pd.DateOffset._cache = {}
off = pd.DateOffset(n=1)
hour = pd.offsets.BusinessHour(n=2)

>>> off._cache is hour._cache
True
```

Reply:
right so we can't use the ``@cache_readonly`` decorator here? (or is that what you are doing by pre-emptively setting the cache)?

Reply:
`cache_readonly.__get__` checks for a `_cache` attribute and creates one if it does not exist.  Creating it in `__init__` ensures that one exists, so a new one does not have to be created (since attempting to do so would raise)

Reply:
i.e. this is necessary in order to retain the existing usages of `cache_readonly` on `DateOffset` subclasses.

Reply:
ok makes sense